flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,267 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
 } // namespace Fortran
 
 namespace {
+namespace looputils {
+
+/// Collects info needed about the indcution/iteration variable for each `do
+/// concurrent` in a loop nest. This includes:
+/// * the operation allocating memory for iteration variable,
+/// * the operation(s) updating the iteration variable with the current
+///   iteration number.
+struct InductionVariableInfo {
+  mlir::Operation *iterVarMemDef;
+  llvm::SetVector<mlir::Operation *> indVarUpdateOps;
+};
+
+using LoopNestToIndVarMap =
+    llvm::MapVector<fir::DoLoopOp, InductionVariableInfo>;
+
+/// Given an operation `op`, this lambda returns true if `op`'s operand is
+/// ultimately the loop's induction variable. Detecting this helps finding the
+/// live-in value corresponding to the induction variable in case the induction
+/// variable is indirectly used in the loop (e.g. throught a cast op).
+bool isIndVarUltimateOperand(mlir::Operation *op, fir::DoLoopOp doLoop) {
+  if (auto storeOp = mlir::dyn_cast_if_present<fir::StoreOp>(op))
```

Comment:
Could you consider putting this inside a templatized helper function? At just two types (`fir::StoreOp` & `fir::ConvertOp`) I understand this templatized helper might not mean too much utility but in case the body of the `if` gets more complex, having it inside a template might preclude duplicating changes.

Reply:
As discussed on Teams, this won't work in this case.